pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/BaseForwardIndexBasedIndexHandler.java

```diff
@@ -0,0 +1,86 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.segment.local.segment.index.loader;
+
+import java.io.IOException;
+import java.util.HashSet;
+import java.util.Set;
+import org.apache.pinot.segment.spi.ColumnMetadata;
+import org.apache.pinot.segment.spi.SegmentMetadata;
+import org.apache.pinot.segment.spi.creator.IndexCreatorProvider;
+import org.apache.pinot.segment.spi.store.ColumnIndexType;
+import org.apache.pinot.segment.spi.store.SegmentDirectory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Base class for all of the {@link IndexHandler} classes which need to build their indexes from the forward index.
```

Comment:
I am a bit confused by this. 

IIUC, a particular IndexHandler is supposed to create that index or remove / cleanup / update that index. 

We recently added `ForwardIndexHandler` as well which can rewrite forward index (for work done in https://github.com/apache/pinot/issues/9348) and also remove forward index for noForwardIndex feature. 

Similarly this PR **uses ForwardIndexHandler to rebuild forward index** using some other structures.

So I am not sure if above statement **"all of the IndexHandler classes which need to build their indexes from the forward index.** is exactly true or intuitive ?

Reply:
There are a list of indexes that require the forward index to build their indexes, such as inverted index, JSON index, etc. For most scenarios within the `ForwardIndexHandler` we too need the forward index to modify the forward index (except when we enable forward index for columns where it's disabled). Does that clarify your question? Yes each `IndexHandler` needs to deal with remove / cleanup / update of its own index, but they often rely on forward index to do so.

I'm trying to abstract away the common logic needed by all such `IndexHandlers` as they all may require the need to create a temp forward index and clean it up later on. I can avoid adding this abstract class but then we'll have to duplicate this logic in all such `IndexHandlers`. Let me know if this is preferred over what I'm doing now.